sympy/core/basic.py

```diff
@@ -1771,10 +1771,10 @@ def rewrite(self, *args, **hints):
             if isinstance(args[-1], string_types):
                 rule = '_eval_rewrite_as_' + args[-1]
             else:
-                try:
-                    rule = '_eval_rewrite_as_' + args[-1].__name__
-                except:
-                    rule = '_eval_rewrite_as_' + args[-1].__class__.__name__
+                name = getattr(args[-1], '__name__', None)
```

Comment:
Is this OK?
```
                name = getattr(
                    args[-1], '__name__', args[-1].__class__.__name__)
```
Or could we use `func_name` from utilities?

Reply:
The difference is that then the default expression is evaluated even if not needed. The performance hit from that is minimal but I guess in principle there could be an object that had `.__name__` but not `.__class__` which would then raise.

Reply:
I left this as it is but I'm happy to change it

Reply:
Oh yes...all args evaluate. OK.